### DISCUSSION TITLE
What does Deepgram expect for 2-channel/multichannel encoding of muLaw?

### DISCUSSION THREAD

**Discussion by JamesKiddSmyth** [deepgram/community#695] [General help]
How does Deepgram expect  2-channel/multichannel audio to be mixed? I am trying to send in two channels of mu-law audio and I am just interleaving each byte.  Frankly, I got the alternating idea from how Genesys AudioHook does it, and this [deepgram public example](https://github.com/deepgram/public-code-examples/blob/master/twilio/python/twilio-proxy-stereo.py).  Is this documented anywhere?
```

ws = Client.connect('wss://api.deepgram.com/v1/listen?encoding=mulaw&sample_rate=8000&endpointing=true&channels=2&multichannel=true', headers = headers)
try:

    dumped = False
    send_size = 1024
    total_size = len(mulawBytes)
    delayCount = 5
    delaySize = delayCount * send_size

    for i in range(0, total_size, send_size):
        chunk = mulawBytes[i:i+send_size]
        
        mixed_chunk = bytearray()

        #faked mixing
        for i in range(0, len(chunk)):
            mixed_chunk.append(chunk[i])

            #fake a second channel
            mixed_chunk.append(chunk[i])
            # below starts from a delay and cycles back through the start just to prove my naive mixing isn't degenerate.
            # if i < delaySize:
            #    mixed_chunk.append(mulawBytes[total_size - delaySize + i])
            #else:
            #    mixed_chunk.append(mulawBytes[i - delaySize])

        send_chunk = bytes(mixed_chunk)

        ws.send(send_chunk)
        response = ws.receive(0)

        if response is not None:
            print("Response:", response)

    response = ws.receive()

    if response is not None:
        print("Response:", response)

    exit()
```

Deepgram response indicates two(2) channels ("channel_index":[0,2]) but is only giving me one(1).  And it looks like it is correctly transcribing one of the channels.  But not giving me the other.

`Response: {"type":"Results","channel_index":[0,2],"duration":1.74,"start":0.0,"is_final":true,"speech_final":true,"channel":{"alternatives":[{"transcript":"mom what's your favorite color","confidence":0.9941406,"words":[{"word":"mom","start":0.4325,"end":0.51113635,"confidence":0.91748047},{"word":"what's","start":0.6684091,"end":0.8256818,"confidence":0.9941406},{"word":"your","start":0.8256818,"end":1.0615909,"confidence":0.9975586},{"word":"favorite","start":1.0615909,"end":1.2975,"confidence":0.9970703},{"word":"color","start":1.2975,"end":1.5334091,"confidence":0.93652344}]}]},"metadata":{"request_id":"f90fc83d-ff1e-4e03-b03f-edf7b2640c68","model_info":{"name":"general","version":"2024-01-26.8851","arch":"base"},"model_uuid":"1ed36bac-f71c-4f3f-a31f-02fd6525c489"},"from_finalize":false}`

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Reply by JamesKiddSmyth** [reply]
Note that this example has only one final ws.receive(), but subsequent calls to receive() are None.  So i don't think the second channel in a separate response.

**Comment by JamesKiddSmyth** [comment]
Im getting the impression that Deepgram won't actually treat the two channels independently and transcribe both of them.  It works fine if the speech doesn't overlap, ie. you get first channel 1 transcript and then channel 2 transcript.  Im not sure to what degree it picks one over the other, but it doesn't seem to want to always do both.  Which is pretty lame actually that you would loose transcription fidelity like this.

**Comment by RyanChimienti** [comment]
You are correct--Deepgram interleaves the two channels exactly the same way as Genesys AudioHook. Based on the fact that your transcript for channel 0 looks good, it's safe to say that you got the mixing logic right. 

Deepgram does treat the two channels independently, so whether the speech is overlapping doesn't matter. In this case, I think the issue is that you're not reading all of the websocket messages from Deepgram. Try looping on `ws.receive()` after you've finished sending the audio to Deepgram, and I bet you'll see the rest of the transcripts.

**Reply by JamesKiddSmyth** [reply]
Let me ask you this.  On the other thread, I talk about this AudioHook emulator which I'm built to 1) test my own server-side implementation (wherein we pass the audio off to Deepgram using the python LiveClient) and 2) exercise the Deepgram Genesys integration endpoint.   That emulator connects via websocket to my server, and listens to the microphone to send audio.  So, my server is effectively just proxying the raw bytes to Deepgram.  But, I'm having some kind of issue where Deepgram won't transcribe mulaw audio.   It works fine if I use Linear16, but fails for mu-law.  I know the mu-law is good because i can send it directly from the emulator to Deepgram via 1) the LiveClient SDK, 2) connecting to the deepgram wss://api.deepgram.com/v1/listen?encoding=mulaw&sample_rate=8000&endpointing=true&channels=2&multichannel=true, and 3) sending it to the Deepgram Genesys Integration.  

And I can't for the life of me figure out why mu-law isn't working.  I've triple checked the config/options, which are obviously different for Linear16, mu-law, 1/2-channels, etc.  I've look at the raw bytes in several ways (ping-pong them back to the emulator and let it do a byte comparison, written to files and verified checksums, plotted it).   I feel like I'm missing something very obvious (which you probably can't help with) or there is something very obscure going on (which you might have a suggestion for).

We also have tons of experience using Deepgram in our product, for what that's worth.

And i have another write-up of the problem here, in case it is any clearer: https://stackoverflow.com/questions/78331485/audiohook-emulator-to-deepgram-liveclient-transcription-challenge

Thanks,
James

**Reply by RyanChimienti** [reply]
Based on everything you've described, it sounds like there's some subtle difference between what you're doing on your server and what you're doing when you send the audio directly from the emulator to `wss://api.deepgram.com/v1/listen?encoding=mulaw&sample_rate=8000&endpointing=true&channels=2&multichannel=true`. 

One thing you could try is to take the LiveClient SDK out of the picture entirely and just talk to that URL on your server. That would avoid any potential issues with differing SDK versions between your server and what you've tested on the emulator. And in my opinion that SDK doesn't add much convenience beyond just opening a websocket anyway.

If you give me a request ID from one of the failed sessions, I can take a peek at it and see if anything stands out. In any case let me know if you learn anything else, now you've got me curious too!

**Reply by JamesKiddSmyth** [reply]
Okay, here is a request_id for what i think should be good audio:  7fe196e9-52fb-45fb-bf5d-e378af9768f6
That same audio was transcribed correctly by the websocket api request_id: 4d27e1bb-0b03-4437-80a9-c07abab6ebd6

**Reply by JamesKiddSmyth** [reply]
Okay, now i think that the SDK doesn't do two channel mulaw well.  Only the websocket endpoint does it correctly for me.  I've include both scripts and results after each below.  NOTE: this is obviously not my emulator->server->deepgram issue.  I was just trying to get a good example of _that_ and ran into this.

```
from asyncio import sleep
import datetime
import sys
import time
import pyaudio
import audioop

from simple_websocket import Client, ConnectionClosed

from deepgram import LiveClient, LiveOptions, ClientOptionsFromEnv, LiveTranscriptionEvents

# more complex example
class MyLiveClient(LiveClient):
    def __init__(self, config: LiveClient):
        super().__init__(config)
        super().on(LiveTranscriptionEvents.Transcript, self.on_message)
        super().on(LiveTranscriptionEvents.Metadata, self.on_metadata)
        super().on(LiveTranscriptionEvents.SpeechStarted, self.on_speech_started)
        super().on(LiveTranscriptionEvents.UtteranceEnd, self.on_utterance_end) 
        super().on(LiveTranscriptionEvents.Error, self.on_error)
        
        self.transcriptionCount = 0
        self.transcripts = []
        self.accumulated_responses = []
        self.be_threads = []
        
        # self.test = "child"

    def on_message(self, parent, result, **kwargs):
        print(f"{datetime.datetime.now()} result: {result}") 

    def on_metadata(self, parent, metadata, **kwargs):
        print(f"\n\n{metadata}\n\n")

    def on_speech_started(self, parent, speech_started, **kwargs):
        print(f"\n\n{speech_started}\n\n")

    def on_utterance_end(self, parent, utterance_end, **kwargs):
        print(f"\n\n{utterance_end}\n\n")

    def on_error(self, parent, error, **kwargs):
        print(f"\n\n{error}\n\n")



seconds = 3

if len(sys.argv) > 1:
    seconds = int(sys.argv[1])

print(f"Seconds:", seconds)

chunk = 1024  # Record in chunks of 1024 samples
sample_format = pyaudio.paInt16  # 16 bits per sample
audio_channels = 1
fs = 8000  # Record at 8000 samples per second
seconds = 3
filename = "output.wav"

p = pyaudio.PyAudio()  # Create an interface to PortAudio

print('Recording')

stream = p.open(format=sample_format,
                channels=audio_channels,
                rate=fs,
                frames_per_buffer=chunk,
                input=True)


linearBytes = bytearray()
mulawBytes = bytearray()

dumped = False
# Store data in chunks for 3 seconds
for i in range(0, int(fs / chunk * seconds)):
    data = stream.read(chunk)
    linearBytes.extend(data)
    chunk_mulaw = audioop.lin2ulaw(data, 2)
    if not dumped:
        #print("Chunk mulaw:", chunk_mulaw) 
        dumped = True
    mulawBytes.extend(chunk_mulaw)

stream.stop_stream()
stream.close()
p.terminate()

print('Finished recording')

liveClient = MyLiveClient(ClientOptionsFromEnv())
options = LiveOptions(
    model="nova-2",
    punctuate=True,
    language="en-US",
    encoding="mulaw",
    sample_rate=8000,
    channels=2,
    vad_events=True,
)

liveClient.start(options, addons={'myattr': "hello"}, test="hello")

for i in range(0, 5):
    print('')

try:

    dumped = False
    send_size = 1024
    total_size = len(mulawBytes)

    for i in range(0, total_size, send_size):
        chunk = mulawBytes[i:i+send_size]
        
        mixed_chunk = bytearray()

        #faked mixing
        for i in range(0, len(chunk)):
            mixed_chunk.append(chunk[i])
            mixed_chunk.append(chunk[i])

        send_chunk = bytes(mixed_chunk)

        liveClient.send(send_chunk)

    while True:
        time.sleep(1)


except (KeyboardInterrupt, EOFError, ConnectionClosed):
    if liveClient is not None:
        liveClient.finish()
    exit()
```

Gives Results 

```
result: {
    "type": "Results",
    "channel_index": [
        0,
        1
    ],
    "duration": 2.64,
    "start": 0.0,
    "is_final": true,
    "speech_final": true,
    "channel": {
        "alternatives": [
            {
                "transcript": "Testing. Testing. Testing.",
                "confidence": 0.9943096,
                "words": [
                    {
                        "word": "testing",
                        "start": 1.04,
                        "end": 1.12,
                        "confidence": 0.89100325,
                        "punctuated_word": "Testing."
                    },
                    {
                        "word": "testing",
                        "start": 1.12,
                        "end": 1.62,
                        "confidence": 0.9989564,
                        "punctuated_word": "Testing."
                    },
                    {
                        "word": "testing",
                        "start": 1.8399999,
                        "end": 2.34,
                        "confidence": 0.9943096,
                        "punctuated_word": "Testing."
                    }
                ]
            }
        ]
    },
    "metadata": {
        "request_id": "592f31db-1808-4107-a9a4-5421482357ce",
        "model_info": {
            "name": "2-general-nova",
            "version": "2024-01-11.36317",
            "arch": "nova-2"
        },
        "model_uuid": "1dbdfb4d-85b2-4659-9831-16b3c76229aa"
    }
}

{
    "type": "Metadata",
    "transaction_key": "deprecated",
    "request_id": "592f31db-1808-4107-a9a4-5421482357ce",
    "sha256": "incomplete",
    "created": "2024-04-16T17:43:02.756Z",
    "duration": 2.64,
    "channels": 1,
    "models": [
        "1dbdfb4d-85b2-4659-9831-16b3c76229aa"
    ],
    "model_info": {
        "1dbdfb4d-85b2-4659-9831-16b3c76229aa": {
            "name": "2-general-nova",
            "version": "2024-01-11.36317",
            "arch": "nova-2"
        }
    }
}
```

But this code 

```
import sys
import time
import pyaudio
import audioop

from simple_websocket import Client, ConnectionClosed

seconds = 3

if len(sys.argv) > 1:
    seconds = int(sys.argv[1])

print(f"Seconds:", seconds)

chunk = 1024  # Record in chunks of 1024 samples
sample_format = pyaudio.paInt16  # 16 bits per sample
audio_channels = 1
fs = 8000  # Record at 8000 samples per second
seconds = 3
filename = "output.wav"

p = pyaudio.PyAudio()  # Create an interface to PortAudio

print('Recording')

stream = p.open(format=sample_format,
                channels=audio_channels,
                rate=fs,
                frames_per_buffer=chunk,
                input=True)


linearBytes = bytearray()
mulawBytes = bytearray()

dumped = False
# Store data in chunks for 3 seconds
for i in range(0, int(fs / chunk * seconds)):
    data = stream.read(chunk)
    linearBytes.extend(data)
    chunk_mulaw = audioop.lin2ulaw(data, 2)
    if not dumped:
        #print("Chunk mulaw:", chunk_mulaw) 
        dumped = True
    mulawBytes.extend(chunk_mulaw)

stream.stop_stream()
stream.close()
p.terminate()

print('Finished recording')

headers = {'Authorization': 'Token xxxxxxxx'}

for i in range(0, 5):
    print('')

ws = Client.connect('wss://[api.deepgram.com/v1/listen?encoding=mulaw&sample_rate=8000&endpointing=true&channels=2&multichannel=true](http://api.deepgram.com/v1/listen?encoding=mulaw&sample_rate=8000&endpointing=true&channels=2&multichannel=true)', headers = headers)
try:

    dumped = False
    send_size = 1024
    total_size = len(mulawBytes)

    for i in range(0, total_size, send_size):
        chunk = mulawBytes[i:i+send_size]
        
        mixed_chunk = bytearray()

        #faked mixing
        for i in range(0, len(chunk)):
            mixed_chunk.append(chunk[i])
            mixed_chunk.append(chunk[i] )

        send_chunk = bytes(mixed_chunk)

        ws.send(send_chunk)

    while True:
        response = ws.receive(1)
        print("Response:", response)
        time.sleep(1)

except (KeyboardInterrupt, EOFError, ConnectionClosed):
    ws.close()
```

Gives this response

```
Response: {"type":"Results","channel_index":[0,2],"duration":2.87,"start":0.0,"is_final":true,"speech_final":true,"channel":{"alternatives":[{"transcript":"testing testing testing","confidence":0.9980469,"words":[{"word":"testing","start":0.59583336,"end":0.9136111,"confidence":0.9951172},{"word":"testing","start":1.1519445,"end":1.5491667,"confidence":0.9980469},{"word":"testing","start":1.7875,"end":2.2641666,"confidence":0.99853516}]}]},"metadata":{"request_id":"49e5d53f-c8ed-4682-9ade-926acc53be18","model_info":{"name":"general","version":"2024-01-26.8851","arch":"base"},"model_uuid":"1ed36bac-f71c-4f3f-a31f-02fd6525c489"},"from_finalize":false}


Response: {"type":"Results","channel_index":[1,2],"duration":2.87,"start":0.0,"is_final":true,"speech_final":true,"channel":{"alternatives":[{"transcript":"testing testing testing","confidence":0.9980469,"words":[{"word":"testing","start":0.59583336,"end":0.9136111,"confidence":0.9946289},{"word":"testing","start":1.1519445,"end":1.5491667,"confidence":0.9980469},{"word":"testing","start":1.7875,"end":2.2641666,"confidence":0.99853516}]}]},"metadata":{"request_id":"49e5d53f-c8ed-4682-9ade-926acc53be18","model_info":{"name":"general","version":"2024-01-26.8851","arch":"base"},"model_uuid":"1ed36bac-f71c-4f3f-a31f-02fd6525c489"},"from_finalize":false}
```

**Reply by JamesKiddSmyth** [reply]
"One thing you could try is to take the LiveClient SDK out of the picture entirely and just talk to that URL on your server."

This works.

**Reply by RyanChimienti** [reply]
Sweet, glad you got it working! I followed up on that request ID out of curiosity (7fe196e9-52fb-45fb-bf5d-e378af9768f6), and I found at least one issue. The request came into this URL: `/v1/listen?channels=2&encoding=mulaw&language=en-US&model=nova-2&punctuate=true&sample_rate=8000&vad_events=true&myattr=hello`, which is missing `multichannel=true`.
